embassy-stm32/src/dcmi.rs

```
@@ -0,0 +1,393 @@
+            _ => {
+                #[cfg(feature = "defmt")]
+                defmt::panic!("Invalid pin configuration.");
+                #[cfg(not(feature = "defmt"))]
+                core::panic!("Invalid pin configuration.");
+            }
+        };
+
+        irq.set_handler(Self::on_interrupt);
+        irq.unpend();
+        irq.enable();
+
+        unsafe {
+            T::regs().cr().modify(|r| {
+                r.set_cm(true); // disable continuous mode (snapshot mode)
+                r.set_ess(use_embedded_synchronization);
+                r.set_pckpol(pixclk_polarity == PixelClockPolarity::RisingEdge);
+                r.set_vspol(vsync_level == VSyncDataInvalidLevel::High);
+                r.set_hspol(hsync_level == HSyncDataInvalidLevel::High);
+                r.set_fcrc(0x00); // capture every frame
+                r.set_edm(edm); // extended data mode
+            });
+            T::regs().ier().modify(|r| {
+                r.set_ovr_ie(true);
+                r.set_frame_ie(true);
+            });
+        }
+
+        Self {
+            phantom: PhantomData,
+            dma,
+        }
+    }
+
+    unsafe fn on_interrupt(_: *mut ()) {
+        let ris = T::regs().ris().read();
+        if ris.err_ris() {
+            #[cfg(feature = "defmt")]
+            defmt::error!("DCMI: Error.");
+        }
+        if ris.ovr_ris() {
+            #[cfg(feature = "defmt")]
+            defmt::error!("DCMI: Overrun.");
+        }
+        if ris.frame_ris() {
+            #[cfg(feature = "defmt")]
+            defmt::error!("DCMI: Frame captured.");
+        }
+        T::regs().icr().write(|r| {
+            r.set_err_isc(true);
+            r.set_ovr_isc(true);
+            r.set_frame_isc(true);
+        });
+    }
+
+    unsafe fn toggle(enable: bool) {
+        T::regs().cr().modify(|r| {
+            r.set_enable(enable);
+            r.set_capture(enable);
+        })
+    }
+
+    pub async fn capture(&mut self, buffer: &mut [u32])
```

This should propagate the errors from the irq above (error/overrun) to the user with a `Result`. I'm not sure what DMA will do on overrun, but it might hang forever or do other bad stuff...